game/scripts/vscripts/heroes/tusk/snowball.lua

```diff
@@ -0,0 +1,362 @@
+  particle = ParticleManager:CreateParticle("particles/units/heroes/hero_tusk/tusk_snowball.vpcf", PATTACH_ABSORIGIN_FOLLOW, self.dummy)
+  ParticleManager:SetParticleControl(particle, 0, caster:GetAbsOrigin())
+  ParticleManager:SetParticleControlEnt(particle, 1, target, PATTACH_POINT_FOLLOW, "attach_hitloc", target:GetAbsOrigin(), true) --Target
+  ParticleManager:SetParticleControl(particle, 2, Vector(0,0,0)) --Velocity
+  ParticleManager:SetParticleControl(particle, 3, Vector(radius,radius,radius)) --Radius
+
+  -- Store to change CP's later
+  self.particle = particle
+  self.unitsHit = {}
+  self.unitInSnowball = {}
+  self.unitInSnowball[caster] = true
+ end
+
+-- Launches the snowball to target
+function tusk_snowball_lua:ReleaseSnowball()
+  local caster = self:GetCaster()
+  local target = self.target
+  local snowball_speed = self:GetSpecialValueFor("snowball_speed")
+  local talent = self:GetCaster():FindAbilityByName("special_bonus_unique_tusk_3_lua")
+  if talent then
+    snowball_speed = snowball_speed + talent:GetSpecialValueFor("value")
+  end
+  ParticleManager:SetParticleControl(self.particle, 2, Vector(snowball_speed,snowball_speed,snowball_speed  )) --Velocity
+
+  caster:EmitSound("Hero_Tusk.Snowball.Loop")
+  local projectile_table =
+  {
+    Target = target,
+    Source = caster,
+    Ability = self,
+    EffectName = "particles/dev/empty_particle.vpcf",
+    iMoveSpeed = snowball_speed,
+    vSourceLoc= caster:GetAbsOrigin(),
+    bDrawsOnMinimap = false,
+    bDodgeable = false,
+    bIsAttack = false,
+    bVisibleToEnemies = true,
+    flExpireTime = GameRules:GetGameTime() + 10,
+    bProvidesVision = false,
+  }
+  ProjectileManager:CreateTrackingProjectile(projectile_table)
+end
+
+---@override
+function tusk_snowball_lua:OnProjectileThink(vLocation)
+  local caster = self:GetCaster()
+  local stun_duration = self:GetSpecialValueFor("stun_duration")
+  local damage = self:GetSpecialValueFor("snowball _damage")
+
+  local talent = self:GetCaster():FindAbilityByName("special_bonus_unique_tusk_2_lua")
+  if talent then
+    damage = damage + talent:GetSpecialValueFor("value")
+  end
+  local damage_per_unit = self:GetSpecialValueFor("snowball_damage_bonus")
+
+  -- Move units along
+  for unit,_ in pairs(self.unitInSnowball) do
+    unit:SetAbsOrigin(vLocation)
+  end
+
+  -- Damage enemy units
+  local units = FindUnitsInRadius(caster:GetTeamNumber(),vLocation,nil,self.radius,DOTA_UNIT_TARGET_TEAM_ENEMY,DOTA_UNIT_TARGET_HERO+DOTA_UNIT_TARGET_BASIC,DOTA_UNIT_TARGET_FLAG_NONE,FIND_ANY_ORDER,false)
+  for _,unit in pairs(units) do
+    if not self.unitsHit[unit] then
+      local damage_table = {
+        victim = unit,
+        attacker = caster,
+        ability = self,
+        damage = damage + (#self.unitInSnowball-1 * damage_per_unit),
+        damage_type = self:GetAbilityDamageType(),
+      }
+      ApplyDamage(damage_table)
+      unit:AddNewModifier(caster,self,"modifier_stunned",{duration = stun_duration})
```

Comment:
Why is this not in OnProjectileHit, would this not apply multiple times to the same units

Reply:
It's a tracking projectile, I would be surprised if it triggers reaching units that are not the target. 
I did indeed forget to self.unitsHit[unit] = true